### DISCUSSION TITLE
How to add JSON type attribute in directive annotation class

### DISCUSSION THREAD

**Discussion by Sandip888831** [ExpediaGroup/graphql-kotlin#1913] [Q&A] [answered]
I am creating a custom directive, I have requirement of adding non primitive types attribute something like below:
```kotlin
@GraphQLDirective
annotation class Test( 
  val name: String,
  val data: JSON,
  val obj: CustomObject
)
```

What I read from the documentation is, annotation class doesn't support custom types.
Is there any way to add JSON & CustomObject type attributes?
Thanks in advance.

**Comment by dariuszkuc** [comment]
Hello 👋 
As per [Kotlin documentation](https://kotlinlang.org/docs/annotations.html#constructors), annotations can only use `primitives / Strings / Classes / Enums / Annotations / Arrays (of any of the above)` parameter types.

`CustomObject` could potentially be represented as another `annotation class` but due to the above limitations you won't be able to store arbitrary JSON data. You could potentially store it in a String representation but you would need to update your coercing logic to handle it accordingly.

**Reply by Sandip888831** [reply]
Thank you for your prompt response. We are fairly new to this library & exploring the possibilities defining the custom directive with complex object types. Your further help will be very much appreciated.

Based on your suggestions, we have the following questions:

1. In relation to the JSON query, could you please assist us with the steps to define the Coercing logic ? Where and how can we define it ? If we define the Coercing logic for parsing a JSON String to JSON object conversion, will it be applied to all Strings?

2.On CustomObject query, I am defining it as annotation & trying to create it's instance something like below:
```
@GraphQLDirective
annotation class Test( 
  val name: String,
  // val data: JSON,
  val obj: CustomObject = CustomObject(10,20)
)

annotation class CustomObject(
    val t1: Int,
    val t2: Int
)
```

But Expedia is throwing below error while application startup:
```
Caused by: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'routerFunctionMapping' defined in class path resource [org/springframework/boot/autoconfigure/web/reactive/WebFluxAutoConfiguration$EnableWebFluxConfiguration.class]: 
Invocation of init method failed; nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'com.expediagroup.graphql.server.spring.SdlRouteConfiguration': Bean instantiation via constructor failed; 
nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.expediagroup.graphql.server.spring.SdlRouteConfiguration$$EnhancerBySpringCGLIB$$60e80522]: Constructor threw exception; 
nested exception is java.lang.ClassCastException: class graphql.schema.GraphQLInterfaceType cannot be cast to class graphql.schema.GraphQLInputType (graphql.schema.GraphQLInterfaceType and graphql.schema.GraphQLInputType are in unnamed module of loader 'app')
\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.initializeBean(AbstractAutowireCapableBeanFactory.java:1804)\n\tat org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:620)
```

What other configuration that I should do in custom schema generator hooks to make it work?

**Reply by Sandip888831** [reply]
@dariuszkuc , could you please assist with the query above? We have been using the Expedia graphql-kotlin library for our graphql project, but we have encountered a problem with a custom directive requirement. We are on a tight deadline and need to quickly determine if there is a solution to this issue. Otherwise, we will have to switch to the DGS framework. However, we would like to avoid moving to another framework since we already have a large codebase on the Expedia graphql-kotlin library.

We kindly request your assistance in resolving this issue as a priority.

Thank you very much in advance!

**Reply by dariuszkuc** [reply]
👋 Haven't tried it, but I believe you should be able to achieve it through `SchemaGeneratorHooks` - i.e. when generating [directive](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/generator/graphql-kotlin-schema-generator/src/main/kotlin/com/expediagroup/graphql/generator/hooks/SchemaGeneratorHooks.kt#L81) you would need to provide your own custom directive type definition which would reference your custom input type (which you will also have to create in a separate `willGenerateGraphQLType` hook). This should look similar to what we do for generating [federated directives](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/generator/graphql-kotlin-federation/src/main/kotlin/com/expediagroup/graphql/generator/federation/FederatedSchemaGeneratorHooks.kt#L235).

While directives can use input objects for their arguments, this is not a common use case so out-of-box handling of this wasn't implemented yet (I guess?). I opened up https://github.com/ExpediaGroup/graphql-kotlin/issues/1917 to track it.

**Reply by kschat** [reply]
Hi @dariuszkuc, thanks for the help. I'm on the GraphQL platform team working with @Sandip888831 to get his subgraph onboarded to our federated GraphQL gateway. For completeness here is our full problem statement:

Currently we have a GraphQL directive definition that is used by all subgraphs to do authorization. The definition of this directive contains input objects _and_ nullable input fields and every subgraph _must_ have the exact same definition. When using the facilities provided by this framework to do codegen, we get a definition that doesn't match the existing authorization directive definition. Most subgraphs do `GraphQL` -> `<native language>` codegen, so we haven't run into this problem before.

I'm going to work with @Sandip888831 on using the hook API to do this. Quick question before I do, though. We found that the annotation class method for defining directives doesn't support nullable input fields. Will the hook API solve this issue, too?

Tangential question: do you have plans to support `GraphQL` -> `Kotlin` codegen? In my experience this is the more flexable option since if there's an issue with the native code you can easily wrap it to get the behavior you desire.

Thanks again for your help!

**Reply by dariuszkuc** [reply, marked as answer]
@kschat TLDR is that you can do any sort of modifications through raw GraphQL objects ([see example impl](https://github.com/dariuszkuc/graphql-kotlin/commit/4f8f95e64bbb86d17458ccb36b61dba7248095d1) of directive with custom input object argument). It is doable.... but it is far from the best user experience. 

You need to:
* manually provide type definition for that directive input object definition
* manually transform your applied directive to correctly process the arguments
* in directive wiring manually parse out the values from inner object

re: NULL/optional arg support for arguments -> see https://github.com/ExpediaGroup/graphql-kotlin/issues/1830. Since you can provide those type definitions, you certainly can mark those fields as nullable.... but we are limited by JVM which does not support it, so if you want to apply them in your code then you have to specify some values for them. As such you would need to provide some values and then in one of the hooks filter out the placeholder values.

--- unsure if this would work ---
While it is messy, it might be possible to also define X versions of given directive with permutations of nullable fields -> if it is just 1 field it might be worth taking a look, e.g. `annotation class Foo(val required: String)` and `annotation class FooBar(val required: String, val optional: Int)`

**Reply by dariuszkuc** [reply]
re: `GraphQL` -> `Kotlin` codegen - AFAIK there are no plans to support it. With both approaches you can manually fix the underlying types (in `graphql-kotlin` you would use the hooks api). In the end you are still bound to the underlying platform limitations. The underlying issue is still how do you represent the types and how do you reference them.

**Reply by dariuszkuc** [reply]
Just thought about another workaround -> instead of creating custom annotation and doing all this extra handling, you could just apply the directive directly using hooks (in `didGenerateGraphQLType`). Unsure if it would be any simpler (maybe you can map inputs better?).

**Reply by Sandip888831** [reply]
Thank you @dariuszkuc for your help.
